### ISSUE TITLE
Textual 0.71 introduces another deadlock in Memray's test suite

### ISSUE THREAD

**Issue by godlygeek** [Textualize/textual#4699]
Steps to reproduce:
```sh
mkdir textual-deadlock
cd textual-deadlock
python3.12 -m venv venv
source venv/bin/activate
git clone https://github.com/bloomberg/memray.git -b v1.13.2
pip install memray -r memray/requirements-test.txt
python -m pytest memray/tests/unit/test_tui_reporter.py -k test_update_thread -x
```

Note the `-b` on the `git clone` call - we're going to work around this deadlock, so for reproducing it you'll need to check out a tag prior to our workaround.

`git bisect` says that the deadlock was introduced by a85b4549428c4ffaccb4c8431bc6dca76e8a1b67 and it still occurs on main.

This may be the same issue as #4696. If so, this is a reliable (albeit not minimal) reproducer for it.

**Comment by willmcgugan** [comment]
Thanks. I can reproduce that.

I'm going to reconsider our locking approach. It's too naive, and any change seems to move the issue elsewhere.

**Comment by willmcgugan** [comment]
@godlygeek The deadlock is fixed in 0.72.0

There are some inconsequential snapshot fails. These should be less frequent in future Textual releases.
